src/anndata/_core/extensions.py

```diff
@@ -0,0 +1,139 @@
+from __future__ import annotations
+
+import inspect
+
+# from collections.abc import Callable
+from pathlib import Path
+from typing import TYPE_CHECKING, Generic, TypeVar
+from warnings import warn
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+from anndata import AnnData
+
+# Based off of the extension framework in Polars
+# https://github.com/pola-rs/polars/blob/main/py-polars/polars/api.py
+
+__all__ = ["register_anndata_namespace"]
+
+
+def find_stacklevel() -> int:
+    """
+    Find the first place in the stack that is not inside AnnData.
+
+    Taken from:
+    https://github.com/pandas-dev/pandas/blob/ab89c53f48df67709a533b6a95ce3d911871a0a8/pandas/util/_exceptions.py#L30-L51
+    """
+    import anndata as ad
+
+    pkg_dir = str(Path(ad.__file__).parent)
+
+    # https://stackoverflow.com/questions/17407119/python-inspect-stack-is-slow
+    frame = inspect.currentframe()
+    n = 0
+    try:
+        while frame:
+            fname = inspect.getfile(frame)
+            if fname.startswith(pkg_dir) or (
+                (qualname := getattr(frame.f_code, "co_qualname", None))
+                # ignore @singledispatch wrappers
+                and qualname.startswith("singledispatch.")
+            ):
+                frame = frame.f_back
+                n += 1
+            else:
+                break
+    finally:
+        # https://docs.python.org/3/library/inspect.html
+        # > Though the cycle detector will catch these, destruction of the frames
+        # > (and local variables) can be made deterministic by removing the cycle
+        # > in a finally clause.
+        del frame
+    return n
+
+
+NS = TypeVar("NS")
+
+
+# Currently, there are no reserved namespaces internally, but if there ever are,
+# this will not allow them to be overridden.
+_reserved_namespaces: set[str] = set.union(*(cls._accessors for cls in (AnnData,)))
```

Comment:
```suggestion
_reserved_namespaces: set[str] = set(AnnData._accessors)
```
why not?

Reply:
That should work as well, in addition I forgot to add all the attributes of AnnData so those don't get overwritten as you've stated [here][comment]. Will add that like so:

```python
_reserved_namespaces: set[str] = set(AnnData._accessors) | set(dir(ad.AnnData))
```

[comment]:
https://github.com/scverse/anndata/pull/1870#discussion_r1967424266